augur/curate/rename.py

```diff
@@ -0,0 +1,85 @@
+"""
+Renames fields / columns of the input data
+"""
+
+from typing import Iterable, Literal, Union, List, Tuple
+import argparse
+from augur.io.print import print_err
+from augur.errors import AugurError
+
+def register_parser(parent_subparsers):
+    parser = parent_subparsers.add_parser("rename",
+    parents = [parent_subparsers.shared_parser],
+    help = __doc__)
+
+    p = parser.add_argument_group(title="RENAME SPECIFIC OPTIONS")
+
+    p.add_argument("--field-map", nargs="+", default=[],
+        help="Rename fields/columns via '{old_field_name}={new_field_name}'. " +
+             "If the new field already exists, then the renaming of the old field will be skipped. " +
+             "Multiple entries with the same '{old_field_name}' will duplicate the field/column. " +
+             "Skips the field if the old field name is the same as the new field name (case-sensitive).")
+    p.add_argument("--force", action="store_true",
+        help="Force renaming of old field even if the new field already exists. " +
+             "Please keep in mind this will overwrite the value of the new field.")
+
+    return parser
+
+
+def parse_field_map(field_map_arg: List[str]) -> List[Tuple[str,str]]:
+    seen_new = set() # keep track of the new field names
+    
+    field_map = []
+    for field in field_map_arg:
+        old_name, new_name = field.split('=')
+
+        if new_name in seen_new:
+            raise AugurError(f"Asked to rename multiple fields to {new_name!r}.")
+        seen_new.add(new_name)
+
+        field_map.append((old_name, new_name))
+    return field_map
+
+
+def transform_columns(existing_fields: List[str], field_map: List[Tuple[str,str]], force: bool) -> List[Tuple[str,str]]:
+    """
+    Calculate the mapping of old column names to new column names
+    """
+    # check that all columns to be renamed exist
+    for idx,names in enumerate(field_map[:]):
```

Comment:
silly python question, why the explicit subscripting instead of just `enumerate(field_map)`?

Reply:
This pattern allows the remove of elements from the original list with `field_map.pop(idx)` below. I believe there's weird behavior if you remove elements from a list while iterating over it.

Reply:
Thanks, suspected something like that.

Reply:
Here's a concrete example:

```py
>>> foo = ['a', 'b', 'c']
>>> for idx,el in enumerate(foo):
...     if el=='b' or el=='c':
...             foo.pop(idx)
... 
'b'
>>> foo 
['a', 'c']
```

You can do clever things like iterating through them in reverse order, but taking a slice is much simpler!

Reply:
A more functional way would be something like

```py
def field_exists(existing_fields: List[str], old_field: str, new_field: str) -> bool:
    """Check if old_field exists in existing_fields, and warn if it doesn't"""
    if old_field in existing_fields:
        return True
    print_err(f"WARNING: Asked to rename field {old_field!r} (to {new_field!r}) but it doesn't exist in the input data.")
    return False


fm_cols_exist = {old_field:new_field for old_field,new_field in field_map.items() if field_exists(existing_fields, old_field, new_field)} 
```

and while I prefer functional style I find it doesn't come naturally in python.